src/main/java/com/google/sps/servlets/utils/Statistics.java

```diff
@@ -15,93 +15,119 @@
    *
-   * @param sentimentScores given score values
+   * @param userCommentList given list of userComment objects
+   * @param topNComments the number of highest magnitudes to retrieve
    */
-  public Statistics(List<Double> sentimentScores) throws RuntimeException {
-    sentimentScores =
-        sentimentScores.stream()
-            .filter(score -> (score >= LOWER_END_VAL && score <= UPPER_END_VAL))
-            .collect(Collectors.toList());
-    setAggregateScores(sentimentScores);
-    setAverageScore(sentimentScores);
+  public Statistics(List<UserComment> userCommentList, int topNComments) {
+    sentimentBucketList = categorizeToBucketList(userCommentList, topNComments);
+    averageScore = getAverageValue(userCommentList, "Average");
+    averageMagnitude = getAverageValue(userCommentList, "Magnitude");
   }
 
   /**
-   * Categorize all score values into different range intervals and count the frequency for each
-   * interval, and set the aggregatedValues.
+   * Categorize all score values into different range SCORE_INTERVALs and count the frequency for
+   * each SCORE_INTERVAL, and set the aggregatedValues.
    *
-   * @param sentimentScores a list of score values from -1.0 to 1.0
+   * @param userCommentList a list of userComment analyzed from sentiment analysis with upadted
+   *     score and magnitude
+   * @return a categorized map based on userCommentList from LOWER_SCORE to UPPER_SCORE with
+   *     SCORE_INTERVAL
    */
-  private void setAggregateScores(List<Double> sentimentScores) {
-    aggregateValues = new HashMap<>();
-    // Add score's interval to different ranges two sorting with and two pointers pop-up
-    sentimentScores.sort(Comparator.naturalOrder());
+  private List<SentimentBucket> categorizeToBucketList(
+      List<UserComment> userCommentList, int topNumComments) {
+    List<SentimentBucket> sentimentBucketList = new ArrayList<>();
+    // Add score's SCORE_INTERVAL to different ranges two sorting with and two pointers pop-up
+    userCommentList.sort(ascendingScoreComparator);
     int updatingScoreIdx = 0;
-    for (BigDecimal tempPoint = LOWER_END;
-        tempPoint.compareTo(UPPER_END) < 0;
-        tempPoint = tempPoint.add(INTERVAL)) {
-      BigDecimal nextPoint = UPPER_END.min(tempPoint.add(INTERVAL));
+    BigDecimal tempPoint;
+    for (tempPoint = LOWER_SCORE;
+        tempPoint.compareTo(UPPER_SCORE) < 0;
+        tempPoint = tempPoint.add(SCORE_INTERVAL)) {
+      BigDecimal nextPoint = UPPER_SCORE.min(tempPoint.add(SCORE_INTERVAL));
       Range currentRange = new Range(tempPoint, nextPoint);
-      aggregateValues.put(currentRange, 0);
+      int currentFrequency = 0;
+      PriorityQueue<UserComment> highMagnitudeComments = new PriorityQueue<>();
       // loop through sorted scores within currentRange from updated score pointer and update its
-      // corresponding appearance frequency in aggregatedValues
+      // corresponding appearance frequency
       int scoreIdx;
```

Comment:
Realized that this variable is unnecessary...
We can simply do:
```
for (updatingScoreIdx = updatingScoreIdx; ...; updatingScoreIdx++)
```
This should save us a variable and a bunch of operations.

Reply:
Cis I guess this is the previous bug we have discussed? scoreIdx is the pointer that iterate through the userCommentList; updatingScoreIdx is the index that represents the start index of the userComment with lowest sentiment score within this range. updatingScoreIdx should be updated only when scoreIdx has reached to the userComment with sentiment score overflowing the current range
Test: try this and the test are failing as expected.